File: app/lib/models/deband_parameters.dart
import 'package:json_annotation/json_annotation.dart';

part 'deband_parameters.g.dart';

/// Parameters for the debanding pass using f3kdb.
/// Removes banding artifacts (color gradients with visible steps).
@JsonSerializable()
class DebandParameters {
  /// Whether this pass is enabled.
  final bool enabled;

  /// Banding detection range (8-128, default 15).
  /// Higher values detect wider bands.
  final int range;

  /// Luma debanding strength (0-64, default 32).
  final int y;

  /// Chroma blue debanding strength (0-64, default 32).
  final int cb;

  /// Chroma red debanding strength (0-64, default 32).
  final int cr;

  /// Dither grain amount (0-64, default 24).
  /// Adds noise to mask remaining banding.
  final int grainY;

  /// Chroma dither grain (0-64, default 24).
  final int grainC;

  /// Use dynamic grain (changes per frame).
  final bool dynamicGrain;

  /// Output bit depth (8, 10, 16).
  final int outputDepth;

  const DebandParameters({
    this.enabled = false,
    this.range = 15,
    this.y = 32,
    this.cb = 32,
    this.cr = 32,
    this.grainY = 24,
    this.grainC = 24,
    this.dynamicGrain = true,
    this.outputDepth = 16,
  });

  DebandParameters copyWith({
    bool? enabled,
    int? range,
    int? y,
    int? cb,
    int? cr,
    int? grainY,
    int? grainC,
    bool? dynamicGrain,
    int? outputDepth,
  }) {
    return DebandParameters(
      enabled: enabled ?? this.enabled,
      range: range ?? this.range,
      y: y ?? this.y,
      cb: cb ?? this.cb,
      cr: cr ?? this.cr,
      grainY: grainY ?? this.grainY,
      grainC: grainC ?? this.grainC,
      dynamicGrain: dynamicGrain ?? this.dynamicGrain,
      outputDepth: outputDepth ?? this.outputDepth,
    );
  }

  /// Get a summary string for display.
  String get summary {
    if (!enabled) return 'Off';
    return 'f3kdb ($y/$cb/$cr)';
  }

  factory DebandParameters.fromJson(Map<String, dynamic> json) =>
      _$DebandParametersFromJson(json);
  Map<String, dynamic> toJson() => _$DebandParametersToJson(this);
}

File: worker/src/models/deband_parameters.rs
use serde::{Deserialize, Serialize};

/// Parameters for the debanding pass using f3kdb.
/// Removes banding artifacts (color gradients with visible steps).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DebandParameters {
    /// Whether this pass is enabled.
    #[serde(default)]
    pub enabled: bool,

    /// Banding detection range (8-128, default 15).
    /// Higher values detect wider bands.
    #[serde(default = "default_range")]
    pub range: i32,

    /// Luma debanding strength (0-64, default 32).
    #[serde(default = "default_y")]
    pub y: i32,

    /// Chroma blue debanding strength (0-64, default 32).
    #[serde(default = "default_cb")]
    pub cb: i32,

    /// Chroma red debanding strength (0-64, default 32).
    #[serde(default = "default_cr")]
    pub cr: i32,

    /// Dither grain amount (0-64, default 24).
    /// Adds noise to mask remaining banding.
    #[serde(default = "default_grain_y")]
    pub grain_y: i32,

    /// Chroma dither grain (0-64, default 24).
    #[serde(default = "default_grain_c")]
    pub grain_c: i32,

    /// Use dynamic grain (changes per frame).
    #[serde(default = "default_dynamic_grain")]
    pub dynamic_grain: bool,

    /// Output bit depth (8, 10, 16).
    #[serde(default = "default_output_depth")]
    pub output_depth: i32,
}

fn default_range() -> i32 { 15 }
fn default_y() -> i32 { 32 }
fn default_cb() -> i32 { 32 }
fn default_cr() -> i32 { 32 }
fn default_grain_y() -> i32 { 24 }
fn default_grain_c() -> i32 { 24 }
fn default_dynamic_grain() -> bool { true }
fn default_output_depth() -> i32 { 16 }

impl Default for DebandParameters {
    fn default() -> Self {
        Self {
            enabled: false,
            range: default_range(),
            y: default_y(),
            cb: default_cb(),
            cr: default_cr(),
            grain_y: default_grain_y(),
            grain_c: default_grain_c(),
            dynamic_grain: default_dynamic_grain(),
            output_depth: default_output_depth(),
        }
    }
}
